## 4. Algorithmic Question

### a) Help Arya by providing a pseudocode for finding an optimal playing strategy, that is, a strategy that maximizes her value. (Hint: Use recursion, assuming that both players play optimally).

#### Pseudocode for maximizeAryaScoreExp(nums, first, last)
**Function**: *maximizeAryaScoreExp(nums, first, last)*

**Input**:  
- *nums*: Array of integers.  
- *first*: The first index of the sequence that Arya can choose from.  
- *last*: The last index of the sequence that Arya can choose from.  

**Output**:  
The maximum score that Arya can achieve, assuming both Arya and Mario play optimally.

&nbsp; **if** *first > last* **then**\
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**return** 0

&nbsp; **if** *first == last* **then** \
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**return** *nums[0]*

&nbsp; **return**  **max** *(*\
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
*nums[first]* + **min** *(* ***maximizeAryaScoreExp*** *(nums, first + 2, last)*, ***maximizeAryaScoreExp*** *(nums, first + 1, last - 1)* *)*,\
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
*nums[last]* + **min** *(*  ***maximizeAryaScoreExp*** *(nums, first + 1, last - 1)*, ***maximizeAryaScoreExp*** *(nums, first, last - 2)* *)*\
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *)*

#### Pseudocode for isAryaWinnerExp(nums)
**Function**: *isAryaWinnerExp(nums)*  

**Input**: *nums*: Array of integers.  

**Output**: A boolean value indicating whether Arya can win assuming both players play optimally.

&nbsp; *n* &larr; length of *nums*\
&nbsp; *aryaScore* &larr; ***maximizeAryaScoreExp*** *(nums, 0, n-1)*\
&nbsp; *marioScore* &larr; **sum**(*nums*) - *aryaScore*\
&nbsp; **if** *aryaScore* $\geq$ *marioScore* **then**\
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**return** *True*\
&nbsp; **else**\
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**return** *False*

#### Correctness of the Algorithm
The algorithm *maximizeAryaScoreExp(nums, first, last)* is correct because it ensures that both players, Arya and Mario, play optimally. By reformulating the problem using **dynamic programming**, it divides the task into subproblems of smaller size, each representing the maximum score Arya can achieve for a given range $[first, last]$ of the sequence. The subproblems are defined by the range $[first, last]$ of the sequence. Each subproblem depends only on smaller subproblems, ensuring that the overlapping subproblem property is satisfied. 

Let *OPT(first, last)* represent the maximum score Arya can achieve if the current range of the sequence is from index *first* to *last*, assuming both players play optimally.

**1. Base Cases**: 
1. If the range of indices is invalid (*first > last*) or *nums* is empty, the algorithm returns $0$ because no numbers are left to pick:
\begin{align*}
OPT(first, last) = 0, \quad \text{if } first > last.
\end{align*}
2. If there is only one number in the range (*first == last*), the algorithm returns that number because Arya has no choice but to pick it:
\begin{align*}
OPT(first, last) = \text{nums}[first], \quad \text{if } first = last
\end{align*}

**2. Recursive Case**: For each recursive call, Arya has two choices:
1. **Pick the first number (*nums[first]*)**: After Arya picks *nums[first]*, Mario chooses optimally to minimize Arya's future score. This is handled by taking the **minimum** of the two possible outcomes: 
     - Arya continues with the range *[first+2, last]* if Mario picks *nums[first+1]*.
     - Arya continues with the range *[first+1, last-1]* if Mario picks *nums[last]*.
   Thus, Arya's score becomes:
   \begin{align*}
   \text{nums}[first] + \min(OPT(first + 2, last), OPT(first + 1, last - 1)).
   \end{align*}

2. **Pick the last number (*nums[last]*)**: After Arya picks *nums[last]*, Mario again minimizes Arya's future score by choosing the best option for himself:
     - Arya continues with the range *[first+1, last-1]* if Mario picks *nums[first]*.
     - Arya continues with the range *[first, last-2]* if Mario picks *nums[last-1]*.
   Similarly to the previous case, Similarly, Arya's score becomes:
   \begin{align*}
   \text{nums}[last] + \min(OPT(first + 1, last - 1), OPT(first, last - 2)).
   \end{align*}

**3. Optimal Strategy**: To ensure that Arya maximizes her score, the algorithm selects the **maximum** of the two possible scores:
1. The score if Arya picks the first number;
2. The score if Arya picks the last number.

Thus, Arya's best score can be formalized as:
\begin{align*}
OPT(first, last) = \max(\text{nums}[first] + \min(OPT(first + 2, last), OPT(first + 1, last - 1)), \text{nums}[last] + \min(OPT(first + 1, last - 1), OPT(first, last - 2))).
\end{align*}

**4. Winning Determination**: Finally, the algorithm *isAryaWinner(nums)* determines the winner by performing the following steps:
1. Computing Arya's score using *maximizeAryaScoreExp(nums, first, last)*;
2. Calculating Mario's score as the difference between the total sum of numbers and Arya's score;
3. Returning *True* if Arya's score is greater or equal to Mario's score, and *False* otherwise.

The correctness of *isAryaWinnerExp(nums)* is directly tied to the correctness of *maximizeAryaScoreExp(nums, first, last)*, as the latter ensures Arya’s score is optimally calculated for any range of numbers.

Therefore, the combination of *maximizeAryaScoreExp* and *isAryaWinnerExp* guarantees correctness because:
- Arya's optimal strategy is computed for every possible range of numbers;
- The comparison in *isAryaWinner* accurately determines the winner based on the scores of Arya and Mario.

### b) Write a Python program implementing her game strategy. Try different array lengths to test the algorithm.

The Python implementation of the algorithms *maximizeAryaScoreExp* and *isAryaWinnerExp* is provided in the `functions.py` module. These are implemented through the functions `maximize_arya_score_exp(nums, first, last)` and `is_arya_winner_exp(nums)`, respectively. 

For a detailed explanation of the functions and their implementation, refer to the `functions.py` module. Below, we present some examples with varying lengths of the array *nums* to test the correctness and performance of the Python program.

In [1]:
from functions import is_arya_winner_exp
# Test with predefined sequences
sequences = [
    [2, 5, 8, 7],
    [1, 5, 2],
    [1, 5, 233, 7],
    [7, 3],
    [5],
    [8, 6, 3, 1, 4],
    [4, 7, 2, 8],
    [1, 15, 15, 10, 3]
]

# Print results for each sequence
for nums in sequences:
    print(f"Sequence: {nums}")
    print(f"Can Arya win?: {is_arya_winner_exp(nums)}")
    print()

Sequence: [2, 5, 8, 7]
Can Arya win?: True

Sequence: [1, 5, 2]
Can Arya win?: False

Sequence: [1, 5, 233, 7]
Can Arya win?: True

Sequence: [7, 3]
Can Arya win?: True

Sequence: [5]
Can Arya win?: True

Sequence: [8, 6, 3, 1, 4]
Can Arya win?: True

Sequence: [4, 7, 2, 8]
Can Arya win?: True

Sequence: [1, 15, 15, 10, 3]
Can Arya win?: False



### c) Is the algorithm efficient? Prove that it is polynomial and provide an asymptotic time complexity bound, or show that it requires exponential time.

The algorithm is inefficient because it requires exponential time. This inefficiency stems from the function *maximizeAryaScoreExp*, which dominates the overall complexity. 

The function *isAryaWinnerExp*, apart from the call to the recursive algorithm *maximizeAryaScoreExp*, has a time complexity of $O(n)$, where $n$ is the length of the input array *nums*. This complexity arises from calculating the total sum of all numbers in *nums* and determining the length of the array. For instance, in Python, the length of a list is stored as an attribute, making this operation $O(1)$. However, in a more general implementation, determining the length may require counting each element, resulting in $O(n)$. Beyond that, *isAryaWinnerExp* performs basic operations such as subtraction and comparison, both of which take constant time $O(1)$. Therefore, the overall time complexity of *isAryaWinnerExp* is dominated by the recursive function *maximizeAryaScoreExp*, which requires exponential time.

Whenever the function *maximizeAryaScoreExp* is called:
1. It performs some comparisons, which take $O(1)$;
2. It makes **4 recursive calls** on inputs of size $n-2$;
3. It computes two minimum values between scores, taking $O(1)$;
4. It performs two additions, taking $O(1)$;
5. It computes a maximum between two scores, taking $O(1)$.

This results in the following recurrence relation:
\begin{aligned}
T(n) = 4 \cdot T(n-2) + c
\end{aligned}
where $c > 0$ is a constant representing the cost of the basic operations.

Let us solve the recurrence relation using the iteration method:

\begin{aligned}
T(n) &= 4 \cdot T(n-2) + c \\
&= 4^1 \cdot T(n-2 \cdot 1) + 4^0 \cdot c \\
&= 4 \cdot (4 \cdot T(n-4) + c) + c \\
&= 4^2 \cdot T(n-2 \cdot 2) + (4^0 + 4^1) \cdot c \\
&= 4^2 \cdot (4 \cdot T(n-6) + c) + 5c \\
&= 4^3 \cdot T(n-2 \cdot 3) + (4^0 + 4^1 + 4^2) \cdot c \\
&\dots \\
&= 4^i \cdot T(n-2 \cdot i) + c \cdot \sum_{k=0}^{i-1} 4^k
\end{aligned}

When $i = \frac{n}{2}$, the recursion reaches the base case $T(0) = 1$. Substituting this value we obtain:

\begin{aligned}
T(n) &= 4^{\frac{n}{2}} \cdot T(0) + c \cdot \sum_{k=0}^{\frac{n}{2}-1} 4^k \\
&= 4^{\frac{n}{2}} + c \cdot \sum_{k=0}^{\frac{n}{2}-1} 4^k
\end{aligned}

The summation $\sum_{k=0}^{\frac{n}{2}-1} 4^k$ is a geometric series:
\begin{aligned}
\sum_{k=0}^{m} 4^k = \frac{4^{m+1} - 1}{4 - 1}
\end{aligned}

Substituting $m = \frac{n}{2} - 1$ we obtain:
\begin{aligned}
\sum_{k=0}^{\frac{n}{2}-1} 4^k = \frac{4^{\frac{n}{2}} - 1}{3}
\end{aligned}

Thus:
\begin{aligned}
T(n) = 4^{\frac{n}{2}} + c \cdot \frac{4^{\frac{n}{2}} - 1}{3} \in O(4^{\frac{n}{2}}) \in o(4^n)
\end{aligned}

Therefore, the algorithm requires exponential time due to the recursive calls in *maximizeAryaScoreExp*, making it impractical for large input sizes.


### d) If the algorithm is exponential, explain how to make it polynomial and provide a pseudocode for it. Recompute the computational complexity of the updated algorithm.

The original exponential algorithm can be transformed into a polynomial-time algorithm by using a matrix to store the results of previously computed subproblems. By doing so, the algorithm avoids redundant calculations, ensuring that each subproblem is solved exactly once. This approach eliminates the need for repeated recursive calls that recompute the same results, which is the primary cause of exponential growth in complexity.

#### Pseudocode for maximizeAryaScorePol(nums)
**Function**: *maximizeAryaScorePol(nums)*

**Input**:  
- *nums*: Array of integers.  

**Output**:  
The maximum score that Arya can achieve, assuming both Arya and Mario play optimally.

&nbsp; *n* &larr; **length**(*nums*)\
&nbsp; *scores* &larr; matrix of size **$n×n$** initialized to 0

&nbsp;&nbsp;&nbsp;&nbsp; **for** *i* **from** 0 **to** *n-1*:  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *scores[i][i]* &larr; *nums[i]*  

&nbsp;&nbsp;&nbsp;&nbsp; **for** *length* **from** 2 **to** *n*:  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; **for** *first* **from** 0 **to** *n* - *length*:  

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *last* &larr; *first* + *length* - 1  

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *choose_first* &larr; *nums[first]* + **min**(  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *scores[first+2][last]* **if** *first*+2 $\geq$ *last* **else** 0,  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *scores[first+1][last-1]* **if** *first*+1 $\geq$ *last*-1 **else** 0)  

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; choose_last &larr; *nums[last]* + **min**(  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *scores[first+1][last-1]* **if** *first*+1 $\geq$ *last*-1 **else** 0,  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *scores[first][last-2]* **if** *first* $\geq$ *last*-2 **else** 0)  

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; *scores[first][last]* &larr; **max**(*choose_first*, *choose_last*)  

&nbsp; **return** *scores[0][n-1]*

#### Pseudocode for isAryaWinnerPol(nums)
**Function**: *isAryaWinnerPol(nums)*  

**Input**: *nums*: Array of integers.  

**Output**: A boolean value indicating whether Arya can win assuming both players play optimally.

&nbsp; *aryaScore* &larr; ***maximizeAryaScorePol*** *(nums)*\
&nbsp; *marioScore* &larr; **sum**(*nums*) - *aryaScore*\
&nbsp; **if** *aryaScore* $\geq$ *marioScore* **then**\
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**return** *True*\
&nbsp; **else**\
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**return** *False*

#### Correctness of the Algorithm

The algorithm *maximizeAryaScorePol(nums)* is correct because it computes the optimal solution by leveraging the same principles as the exponential algorithm *maximizeAryaScoreExp(nums)*, which has already been shown to be correct. Specifically, it ensures correctness by systematically building the solution for the entire problem from the solutions to its subproblems, leveraging **dynamic programming** to iteratively compute the optimal score for progressively larger ranges. This iterative approach guarantees that, when solving a subproblem for the range $[first,last]$, the optimal solutions for all smaller subproblems ($[first+2, last]$, $[first+1, last-1]$, $[first, last-2]$) have already been computed and stored in the *scores* table, ultimately leading to the correct computation of $OPT[0, n-1]$.

This ensures that the algorithm uses the correct values for smaller subproblems when calculating the optimal solution for larger ranges. By mirroring the recursive structure of the exponential algorithm and adhering to the same minimization and maximization principles, the computed results for each range are guaranteed to be optimal. Iterating over subproblem sizes in increasing order ensures every required subproblem is solved before being used, avoiding redundancy and culminating in $OPT[0,n-1]$, the solution to the original problem.

The correctness of *isAryaWinnerPol(nums)* follows directly from the correctness of *maximizeAryaScorePol(nums)*, using the same reasoning as in the exponential case.

#### Complexity Analysis

The overall complexity of *isAryaWinnerPol(nums)* is primarily determined by the time complexity of the *maximizeAryaScorePol(nums)* function. Let us analyze this step by step:
- Calculating the total sum of all numbers in *nums* takes $O(n)$, where $n$ is the length of *nums*;
- Subtraction and comparison operations are constant time $O(1)$.

Thus, the total time complexity for *isAryaWinnerPol(nums)* can be expressed as:
\begin{align*}
T(n) + O(n) + O(1) = T(n) + O(n),
\end{align*}
where $T(n)$ is the time complexity of *maximizeAryaScorePol(nums)*.

Now, let us analyze the complexity of *maximizeAryaScorePol(nums)* in detail:

1. **Initialization**:
   - Computing the length of *nums* takes $O(n)$, as explained in the analysis for the exponential algorithm;
   - Creating an $n \times n$ matrix initialized with zeros takes $O(n^2)$;
   - The first loop that sets the diagonal entries $scores[i][i] = nums[i]$ also takes $O(n)$.

2. **Nested Loops for Dynamic Programming Table Filling**: The nested loops iteratively fill the dynamic programming table. In particular:
     - The **outer loop** iterates over lengths of subarrays from 2 to $n$, contributing $O(n)$;
     - For each length, the **inner loop** iterates over starting indices of subarrays, contributing $O(n - \text{length})$ for each length.
     - **Per-Iteration Operations**, i.e., accessing values in the matrix and performing comparisons and arithmetic operations, take $O(1)$.

     For a fixed value of $length$, the inner loop performs $O(n - \text{length})$ iterations. As $length$ varies from 2 to $n$, the total number of iterations across all loops can be computed as:
     \begin{align*}
     \sum_{\text{length}=2}^{n} \sum_{\text{first}=0}^{n-\text{length}} O(1) = O(n^2).
     \end{align*}

3. **Final Return**: Returning $scores[0][n-1]$ takes $O(1)$.

Combining all the components, we obtain:
\begin{align*}
T(n) = O(n^2) + O(n) + O(n^2) + O(1) \in O(n^2).
\end{align*}

Thus, the total time complexity of *maximizeAryaScorePol(nums)* is $O(n^2)$. Including the $O(n)$ cost from *isAryaWinnerPol(nums)*, the overall complexity remains $O(n^2)$. The space complexity is also $O(n^2)$ due to the $n \times n$ *scores* matrix* used to store intermediate results.

### e) Implement the algorithm in Python. Compare your result values with the previous algorithm. Also compare the running times.

As in the previous case, the Python implementation of the algorithms *maximizeAryaScorePol* and *isAryaWinnerPol* is provided in the `functions.py` module by the functions `maximize_arya_score_pol` and `is_arya_winner_pol`, respectively. These functions include detailed comments and descriptions to ensure clarity and understanding of their purpose and functionality.

Let us now test the Python program using various values and lengths for the *nums* list to verify its correctness and performance. To compare the results with the previous exponential-time implementation, we will use the same inputs previously tested for the *nums* list.

In [2]:
from functions import is_arya_winner_pol
# Test with predefined sequences
sequences = [
    [2, 5, 8, 7],
    [1, 5, 2],
    [1, 5, 233, 7],
    [7, 3],
    [5],
    [8, 6, 3, 1, 4],
    [4, 7, 2, 8],
    [1, 15, 15, 10, 3]
]

# Print results for each sequence
for nums in sequences:
    print(f"Sequence: {nums}")
    print(f"Can Arya win?: {is_arya_winner_pol(nums)}")
    print()


Sequence: [2, 5, 8, 7]
Can Arya win?: True

Sequence: [1, 5, 2]
Can Arya win?: False

Sequence: [1, 5, 233, 7]
Can Arya win?: True

Sequence: [7, 3]
Can Arya win?: True

Sequence: [5]
Can Arya win?: True

Sequence: [8, 6, 3, 1, 4]
Can Arya win?: True

Sequence: [4, 7, 2, 8]
Can Arya win?: True

Sequence: [1, 15, 15, 10, 3]
Can Arya win?: False



Let us summarize the results to compare the outputs of the two algorithms:

| **Sequence**       | **isAryaWinnerExp** | **isAryaWinnerPol** |
|---------------------|---------------------|---------------------|
| [2, 5, 8, 7]        | True                | True                |
| [1, 5, 2]           | False               | False               |
| [1, 5, 233, 7]      | True                | True                |
| [7, 3]              | True                | True                |
| [5]                 | True                | True                |
| [8, 6, 3, 1, 4]     | True                | True                |
| [4, 7, 2, 8]        | True                | True                |
| [1, 15, 15, 10, 3]  | False               | False               |

As we can see, the results are identical for both algorithms. This confirms that the two implementations compute the same optimal strategy for Arya.

To gain further insights, let us test the two algorithms with sequences of varying lengths and diversity. Additionally, we will measure the running times for each algorithm on each sequence to compare their performance. 

Based on our complexity analysis:
- The recursive algorithm *isAryaWinnerExp* has a time complexity of $O(4^{\frac{n}{2}})$, which is exponential;
- The iterative algorithm with memoization *isAryaWinnerPol* has a time complexity of $O(n^2)$, which is polynomial.

Given this exponential gap, we expect the memoization approach to perform significantly better for longer sequences. Let us verify this through empirical testing.


In [4]:
from time import time

sequences = [
    [10, 20, 30, 40, 50, 60],
    [2, 3, 5, 8, 13, 21, 34],
    [10] * 30,
    [1, 100, 1, 100, 1, 100, 1, 100, 1, 100, 1, 1],
    list(range(1, 21)),
    list(range(30, 0, -1)),
    [1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 50, 2, 1, 2, 1],
    [10, 20, 5, 5, 20, 10, 30, 40, 15, 10, 5],
    [1, 10, 20, 8, 5, 5, 10, 20, 10, 30, 40, 50, 60, 15, 10, 5, 1]
]

In [ ]:
# Print results and execution time for each sequence using the exponential time complexity implementation
for nums in sequences:
    start_time = time()
    result = is_arya_winner_exp(nums)
    end_time = time()
    execution_time = end_time - start_time

    print(f"Sequence: {nums}")
    print(f"Can Arya win?: {result}")
    print(f"Execution time: {execution_time:.6f} seconds")
    print()


Sequence: [10, 20, 30, 40, 50, 60]
Can Arya win?: True
Execution time: 0.000000 seconds

Sequence: [2, 3, 5, 8, 13, 21, 34]
Can Arya win?: True
Execution time: 0.000000 seconds

Sequence: [10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10]
Can Arya win?: True
Execution time: 252.753471 seconds

Sequence: [1, 100, 1, 100, 1, 100, 1, 100, 1, 100, 1, 1]
Can Arya win?: True
Execution time: 0.001011 seconds

Sequence: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]
Can Arya win?: True
Execution time: 0.241367 seconds

Sequence: [30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
Can Arya win?: True
Execution time: 244.402064 seconds

Sequence: [1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 50, 2, 1, 2, 1]
Can Arya win?: False
Execution time: 0.005018 seconds

Sequence: [10, 20, 5, 5, 20, 10, 30, 40, 15, 10, 5]
Can Arya win?: True
Execution time: 0.000994 s

In [5]:
# Print results and execution time for each sequence using the polinomial time complexity implementation
for nums in sequences:
    start_time = time()
    result = is_arya_winner_pol(nums)
    end_time = time()
    execution_time = end_time - start_time

    print(f"Sequence: {nums}")
    print(f"Can Arya win?: {result}")
    print(f"Execution time: {execution_time:.6f} seconds")
    print()


Sequence: [10, 20, 30, 40, 50, 60]
Can Arya win?: True
Execution time: 0.000941 seconds

Sequence: [2, 3, 5, 8, 13, 21, 34]
Can Arya win?: True
Execution time: 0.000000 seconds

Sequence: [10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10]
Can Arya win?: True
Execution time: 0.000000 seconds

Sequence: [1, 100, 1, 100, 1, 100, 1, 100, 1, 100, 1, 1]
Can Arya win?: True
Execution time: 0.000000 seconds

Sequence: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]
Can Arya win?: True
Execution time: 0.001003 seconds

Sequence: [30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
Can Arya win?: True
Execution time: 0.000000 seconds

Sequence: [1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 50, 2, 1, 2, 1]
Can Arya win?: False
Execution time: 0.000000 seconds

Sequence: [10, 20, 5, 5, 20, 10, 30, 40, 15, 10, 5]
Can Arya win?: True
Execution time: 0.000000 secon

The following table summarizes the results and execution times for both the exponential algorithm (*isAryaWinnerExp*) and the polynomial algorithm (*isAryaWinnerPol*). Each sequence is tested to determine if Arya can win, along with the respective running times for both algorithms.

| **Sequence**                                                                            | **isAryaWinnerExp**  | **Time (Exp)**    | **isAryaWinnerPol**  | **Time (Pol)**    |
|-----------------------------------------------------------------------------------------|----------------------|-------------------|----------------------|-------------------|
| [10, 20, 30, 40, 50, 60]                                                                | True                 | 0.000000 seconds  | True                 | 0.000000 seconds  |
| [2, 3, 5, 8, 13, 21, 34]                                                                | True                 | 0.000000 seconds  | True                 | 0.000000 seconds  |
| [10]*30                                                                                 | True                 | 252.753471 seconds| True                 | 0.001000 seconds  |
| [1, 100, 1, 100, 1, 100, 1, 100, 1, 100, 1, 1]                                          | True                 | 0.001011 seconds  | True                 | 0.000000 seconds  |
| list(range(1, 21))                                                                      | True                 | 0.241367 seconds  | True                 | 0.000000 seconds  |
| list(range(30, 0, -1))                                                                  | True                 | 244.402064 seconds| True                 | 0.001189 seconds  |
| [1, 2, 1, 2, ..., 50, 2, 1, 2, 1]                                                       | False                | 0.005018 seconds  | False                | 0.000000 seconds  |
| [10, 20, 5, 5, 20, 10, 30, 40, 15, 10, 5]                                               | True                 | 0.000994 seconds  | True                 | 0.000000 seconds  |
| [1, 10, 20, 8, 5, 5, 10, 20, 10, 30, 40, 50, 60, 15, 10, 5, 1]                          | False                | 0.016009 seconds  | False                | 0.000000 seconds  |

As we can see from the table, the results of both algorithms are identical for all test cases, confirming that they implement the same optimal strategy for Arya. However, the execution times reveal significant differences. The exponential algorithm starts to struggle as the input sequence length increases, with execution times growing dramatically. For example, a sequence of length $30$ takes over $\mathbf{252}$ **seconds**, making the exponential approach impractical for inputs of this size or larger. In contrast, the polynomial algorithm handles the same sequence in just $\mathbf{0.001}$ **seconds**, maintaining consistent performance even for significantly longer inputs.

This behavior aligns with the complexity analyses of the two algorithms. The exponential algorithm has a time complexity of $O(4^{n/2})$, which grows extremely quickly as the length of the sequence increases. The polynomial algorithm, on the other hand, has a complexity of $O(n^2)$, allowing it to scale efficiently with the input size. This exponential gap is clearly reflected in the measured execution times.

Overall, these results highlight the practical advantages of the polynomial algorithm. By leveraging memoization, it achieves a drastic reduction in computation time without compromising correctness. For small sequences, both algorithms perform similarly, but for longer sequences, the polynomial approach is vastly superior, making it the clear choice for real-world applications where input sizes may be large.


#### Important Notice

Please note that the results presented in the table may differ slightly from those returned by the code. This is because the code was run multiple times after the initial testing, and in some cases, caching was not disabled. However, this does not affect the overall analysis, as the execution times consistently align with the figures provided in the table and comments. The key takeaway remains the same: the polynomial algorithm significantly outperforms the exponential one for longer sequences while maintaining correctness.

### f) Finally, consult LLM (ChatGPT, Claude AI, Gemini, Perplexity, etc.) to craft a third, optimized implementation and analyze its time complexity. Also, explain if the LLM is doing a good job and how you can evaluate whether the suggested solution works properly.

We asked ChatGPT to write the pseudocode for the problem in a Markdown code cell, using the exercise prompt as input. Below is the pseudocode provided by ChatGPT, which ensures that both Arya and Mario play optimally. It is pasted exactly as it appeared in ChatGPT's output:

```plaintext
Function PredictWinner(nums):
    n ← length(nums)

    # Initialize DP table
    dp ← array of size [n][n] filled with 0

    # Base case: Single elements
    For i from 0 to n-1:
        dp[i][i] ← nums[i]

    # Fill DP table for increasing sequence lengths
    For length from 2 to n:  # Length of the current subarray
        For i from 0 to n-length:
            j ← i + length - 1  # Calculate the end of the subarray
            # Optimal choice for Arya
            dp[i][j] ← max(nums[i] - dp[i+1][j], nums[j] - dp[i][j-1])

    # Arya wins if the score difference she can guarantee is >= 0
    Return dp[0][n-1] >= 0
```

This algorithm is similar to our polynomial algorithm, and therefore the complexity analysis is the same. While this implementation performs fewer constant-time operations within the inner loop and utilizes a single function, the overall analysis remains unchanged since we are concerned with time complexity up to additive constants. As such, the time complexity is $O(n^2)$, and the space complexity is also $O(n^2)$ due to the use of the $n \times n$ DP table.

The correctness of the algorithm stems from its ability to compute the maximum score difference Arya can guarantee for any subarray, assuming both Arya and Mario play optimally. This is achieved using a DP table, where $dp[i][j]$ represents the maximum score difference Arya can achieve over the subarray $nums[i:j+1]$. For single-element subarrays, Arya can only pick the single element, so the value of $dp[i][i]$ is simply $nums[i]$. For larger subarrays, the algorithm recursively computes the maximum score difference based on Arya's two possible moves: picking the first element, resulting in $nums[i] - dp[i+1][j]$, or picking the last element, resulting in $nums[j] - dp[i][j-1]$. The algorithm ensures Arya selects the move that maximizes her guaranteed score difference by taking the maximum of these two values.

The DP table is constructed iteratively, starting with the smallest subarrays and progressively building up to larger ones. By the time the algorithm computes $dp[0][n-1]$, all smaller subproblems have already been solved, ensuring the correctness of the result. The final value of $dp[0][n-1]$ represents the maximum score difference Arya can guarantee for the entire array. If this value is non-negative, Arya can ensure her score is at least equal to Mario's, and the algorithm returns *True*, otherwise it returns *False*.

Let us implement it to check its results.

In [7]:
from functions import predictWinnerChatGPT

sequences = [
    [2, 5, 8, 7],
    [1, 5, 2],
    [1, 5, 233, 7],
    [7, 3],
    [5],
    [8, 6, 3, 1, 4],
    [4, 7, 2, 8],
    [1, 15, 15, 10, 3]
]

# Print results for each sequence
for nums in sequences:
    print(f"Sequence: {nums}")
    print(f"Can Arya win?: {predictWinnerChatGPT(nums)}")
    print()

Sequence: [2, 5, 8, 7]
Can Arya win?: True

Sequence: [1, 5, 2]
Can Arya win?: False

Sequence: [1, 5, 233, 7]
Can Arya win?: True

Sequence: [7, 3]
Can Arya win?: True

Sequence: [5]
Can Arya win?: True

Sequence: [8, 6, 3, 1, 4]
Can Arya win?: True

Sequence: [4, 7, 2, 8]
Can Arya win?: True

Sequence: [1, 15, 15, 10, 3]
Can Arya win?: False



As we stated in the correctness analysis, the algorithm returns the same results as those achieved by our exponential and polynomial algorithms, affirming its correctness. This implementation performs fewer comparisons compared to our polynomial algorithm, as it avoids unnecessary operations. However, since these are constant-time operations, the difference in execution time is imperceptible.

For completeness, let us test this implementation using the same sequences that were used for computing the running times of the previous algorithms. This will provide evidence of its efficiency and demonstrate that its running times differ only marginally from those of our polynomial implementation.

In [8]:
sequences = [
    [10, 20, 30, 40, 50, 60],
    [2, 3, 5, 8, 13, 21, 34],
    [10] * 30,
    [1, 100, 1, 100, 1, 100, 1, 100, 1, 100, 1, 1],
    list(range(1, 21)),
    list(range(30, 0, -1)),
    [1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 50, 2, 1, 2, 1],
    [10, 20, 5, 5, 20, 10, 30, 40, 15, 10, 5],
    [1, 10, 20, 8, 5, 5, 10, 20, 10, 30, 40, 50, 60, 15, 10, 5, 1]
]

# Print results and execution time for each sequence using ChatGPT's polynomial time complexity implementation
for nums in sequences:
    start_time = time()
    result = predictWinnerChatGPT(nums)
    end_time = time()
    execution_time = end_time - start_time

    print(f"Sequence: {nums}")
    print(f"Can Arya win?: {result}")
    print(f"Execution time: {execution_time:.6f} seconds")
    print()


Sequence: [10, 20, 30, 40, 50, 60]
Can Arya win?: True
Execution time: 0.001009 seconds

Sequence: [2, 3, 5, 8, 13, 21, 34]
Can Arya win?: True
Execution time: 0.000000 seconds

Sequence: [10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10]
Can Arya win?: True
Execution time: 0.000000 seconds

Sequence: [1, 100, 1, 100, 1, 100, 1, 100, 1, 100, 1, 1]
Can Arya win?: True
Execution time: 0.000000 seconds

Sequence: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]
Can Arya win?: True
Execution time: 0.000000 seconds

Sequence: [30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
Can Arya win?: True
Execution time: 0.000000 seconds

Sequence: [1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 50, 2, 1, 2, 1]
Can Arya win?: False
Execution time: 0.000000 seconds

Sequence: [10, 20, 5, 5, 20, 10, 30, 40, 15, 10, 5]
Can Arya win?: True
Execution time: 0.000000 secon